### PR TITLE
IFC 42 - Refactor how DiffRepository is stored in the graph

### PR DESCRIPTION
**Pull request by dgarros**
This PR is a refactoring of how a `DiffRepository` is stored internally to query the nodes starting from the children instead of querying from the top of the tree every time.

The need for this change came up while working on #4107 because it was not possible to implement the filters in the current state, if the objects we were interested about were not at the top of the tree, it wouldn't be possible to retrieve them.

The proposed solution is to query all nodes and retrieve the paths to all parents, if any,
In this scenario, the query is actually a lot simpler because we don't need to query the attributes and the relationships for the parents ... 

Related to this change, I've also slightly changed the format of the data returned by the GraphQL query
Instead of having a list of nodes_uuid at the relationship level, I've added a new `parent_node` attribute at the `DiffNode` level, which in return made it possible to hide the relationship from the client if there is no other information on it.

### CODE REVIEW

File: backend/infrahub/graphql/queries/diff/tree.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ class DiffRelationship(DiffSummaryCounts):
     last_changed_at = DateTime(required=False)
     status = Field(GrapheneDiffActionEnum, required=True)
     elements = List(DiffSingleRelationship, required=True)
-    node_uuids = List(String, required=True)
+    node_uuids = List(String, required=True)  # NOTE Can we remove this one now ?
```
Comment (ajtmccarty) [inline backend/infrahub/graphql/queries/diff/tree.py]: yes we can

File: backend/infrahub/core/diff/model/path.py [hunk 2]
```diff
@@ -188,6 +194,12 @@ class EnrichedDiffNode(BaseSummary):
     def __hash__(self) -> int:
         return hash(self.uuid)
 
+    def get_parent_node(self) -> EnrichedDiffNode | None:
+        for r in self.relationships:
+            for n in r.nodes:
+                return n
+        return None
```
Comment (ajtmccarty) [inline backend/infrahub/core/diff/model/path.py:201]: maybe `EnrichedDiffRelationship.nodes` should become `parent_node: EnrichedDiffNode | None` now?
but we can also leave this as is

File: backend/infrahub/graphql/queries/diff/tree.py [hunk 3]
```diff
@@ -100,6 +100,7 @@ class DiffNode(DiffSummaryCounts):
     conflict = Field(ConflictDetails, required=False)
     contains_conflict = Boolean(required=True)
     last_changed_at = DateTime(required=False)
+    parent_node = String(required=False)
```
Comment (ajtmccarty) [inline backend/infrahub/graphql/queries/diff/tree.py:102]: is this enough information for the front-end? I would think that they would need the parent relationship name/label as well, but I'm not sure

**Review by ajtmccarty — APPROVED** [review]

File: backend/infrahub/graphql/queries/diff/tree.py (hunk 3 above)
Comment (dgarros) [reply backend/infrahub/graphql/queries/diff/tree.py:102]: I'll merge as it for now but I'll check with the frontend team tomorrow morning

**Pull request merged**